### PR TITLE
Public group chat (#628)

### LINKED ISSUE #628: Whisper topics as 'public group chats'
We should allow whisper topics to be public group chats anyone can join. The issue here is that it should be made obvious that they are not encrypted, as per current Whisper protocol design.

https://instabug.com/applications/status-10bb8cbc-15f6-44f0-8c7a-db478dc34d7d/beta/bugs/77

### PR DESCRIPTION
**Pull request by rasom**
resolves #628

### CODE REVIEW

**Comment by rasom** [comment]
@jarradh, I have a few question regarding public chat functionality:

1.  Should we have another icon for public groups?


2. I suppose `topic` field is enough for joining public group, do we need to add something else here?
screen shot 2017-02-07 at 13 03 21

3. We don't really know how many users are there, so maybe it would be better to write `available` in status bar or something like this


4. Looks like we need to show usernames in the chat because otherwise it is hard to figure who wrote the message. Also, it is not clear how to handle messages delivery status. 


5. We need to show that the group is public, currently there is no visual difference for user

**Comment by rasom** [comment]
Also, we have a problem with the order of messages when a new user joins the chat:

o

**Comment by jarradh** [comment]
> 1. Should we have another icon for public groups?
Probably a good idea, like a # symbol ? @andmironov  what do you think?

> 2. I suppose topic field is enough for joining public group, do we need to add something else here?
It's enough for now I think

> 3. We don't really know how many users are there, so maybe it would be better to write available in status bar or something like this
Agreed maybe "Public"

> 4. Looks like we need to show usernames in the chat because otherwise it is hard to figure who wrote the message. Also, it is not clear how to handle messages delivery status.
good point @andmironov any ideas?

> 5. We need to show that the group is public, currently there is no visual difference for users
Yes, first I think the icon should be similar to the # symbol icon in 1.
Instead of showing users it should say "Public" as well

> 6. we have a problem with the order of messages
yes.. yes we do.

**Comment by rasom** [comment]
@jarradh, how will we handle delivery here, and how to show delivery status? As an option, we could just set longer ttl for public chats and stop sending the message after the first ack, or even don't wait for ack at all. And, I think `sent` status is enough for this kind of chat. Other ideas?

**Comment by jarradh** [comment]
I agree, I think waiting for ack in this scenario as it may involve many clients, it may become impractical, as long as it is only affecting public group chat. Long TTL is fine.

As for `sent` status I agree.

We'll probably revisit this as a better solution comes to light.

**Comment by rasom** [comment]
thanks for the answer, @jarradh

File: src/status_im/chat/handlers/send_message.cljs [hunk 1]
```diff
@@ -303,10 +317,19 @@
...
                                  :payload    payload}}]
-          (if group-chat
+          (cond
+            (and group-chat (not public?))
             (protocol/send-group-message! (assoc options
                                             :group-id chat-id
                                             :keypair {:public  public-key
                                                       :private private-key}))
+
+            (and group-chat not)
```
Comment (alwx) [inline src/status_im/chat/handlers/send_message.cljs]: `not` looks weird here

File: src/status_im/chat/views/actions.cljs [hunk 2]
```diff
@@ -114,12 +114,12 @@
         subtitle])]]])
 
 (defn actions-list-view []
```
Comment (alwx) [inline src/status_im/chat/views/actions.cljs]: `actions-list-view` doesn't conform re-frame's form 2
maybe it is better to replace it with `defview`

File: src/status_im/chat/views/toolbar_content.cljs [hunk 3]
```diff
@@ -49,38 +49,51 @@
...
+                :font            :toolbar-title}
+          (if @public?
+            (str "#" chat-name)
+            chat-name)])
        (if @group-chat
          [group-last-activity {:contacts   @contacts
+                               :public?    @public?
                                :sync-state @sync-state}]
          [last-activity {:online-text (online-text @contact @chat-id)
                          :sync-state  @sync-state}])])))
```
Comment (alwx) [inline src/status_im/chat/views/toolbar_content.cljs]: `defview` will also look nice here, but, of course, it is neither obligatory nor important

File: src/status_im/contacts/views/contact_list.cljs [hunk 4]
```diff
@@ -21,24 +21,35 @@
...
   (fn [row _ _]
     (list-item
-     [contact-view {:contact  row
-                    :letter?  chat-modal
-                    :on-click (if click-handler
-                                #(click-handler row action params))}])))
+      [contact-view {:contact  row
+                     :letter?  chat-modal
+                     :on-click (if click-handler
+                                 #(click-handler row action params))}])))
```
Comment (alwx) [inline src/status_im/contacts/views/contact_list.cljs]: `if` => `when`

File: src/status_im/data_store/realm/schemas/account/v4/core.cljs [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+             kv-store/schema
+             message/schema
+             pending-message/schema
+             processed-message/schema
+             request/schema
+             tag/schema
+             user-status/schema])
+
+(defn migration [old-realm new-realm]
+  (log/debug "migrating v3 account database: " old-realm new-realm)
```
Comment (alwx) [inline src/status_im/data_store/realm/schemas/account/v4/core.cljs]: v4 :)

File: src/status_im/new_group/handlers.cljs [hunk 6]
```diff
@@ -91,6 +93,43 @@
...
+  (u/side-effect!
+    (fn [db [_ topic]]
+      (let [exists? (boolean (get-in db [:chats topic]))
+            group   {:chat-id     topic
+                     :name        topic
+                     :color       default-chat-color
+                     :group-chat  true
+                     :public?     true
+                     :is-active   true
+                     :timestamp   (.getTime (js/Date.))}]
```
Comment (alwx) [inline src/status_im/new_group/handlers.cljs]: There is a function `timestamp` in `status-im.utils.random` which does the same.
But:
1) I am not sure that `status-im.utils.random` ns is a great place for this function;
2) There are probably other calls to js/Date in our code.

So probably we should create another task for beginners to remove these calls and think about better code organization.

File: src/status_im/new_group/handlers.cljs [hunk 7]
```diff
@@ -91,6 +93,43 @@
...
+          (dispatch [::start-watching-group topic]))))))
+
+(register-handler ::add-public-group
+  (fn [db [_ {:keys [chat-id] :as group}]]
+    (assoc-in db [:chats chat-id] group)))
+
+(register-handler ::save-public-group
+  (u/side-effect!
+    (fn [_ [_ group]]
+      (log/debug :GGG group)
```
Comment (alwx) [inline src/status_im/new_group/handlers.cljs]: Should better be removed

File: src/status_im/protocol/group.cljs [hunk 8]
```diff
@@ -6,22 +6,25 @@
...
+                             :username username
                              :group-id group-id
                              :type type
                              :timestamp (u/timestamp))
                      (assoc :topics [group-id]
-                            :requires-ack? true
-                            :keypair keypair
+                            :requires-ack? (if (nil? requires-ack?)
+                                             true
+                                             requires-ack?)
```
Comment (alwx) [inline src/status_im/protocol/group.cljs]: `(or (nil? requires-ack?) requires-ack?)`

File: src/status_im/chat/views/actions.cljs (hunk 2 above)
Comment (rasom) [inline src/status_im/chat/views/actions.cljs]: Let's add `defview` next time, I just fixed form-2 here

File: src/status_im/chat/views/toolbar_content.cljs (hunk 3 above)
Comment (rasom) [inline src/status_im/chat/views/toolbar_content.cljs]: The same about `defview` here

File: src/status_im/contacts/views/contact_list.cljs (hunk 4 above)
Comment (rasom) [inline src/status_im/contacts/views/contact_list.cljs]: Interesting, in Intellij I even don't see that I've touched this code

**Comment by rasom** [comment]
Thanks for comments @alwx, fixed.

**Comment by jarradh** [comment]
hmm I think we can enforce # as first character in public chat input and not allow spaces to be typed (keep regex in background)

**Comment by rasom** [comment]
I'm not sure how to disallow typing of spaces without flickering of text. Enforcing "#" as the first character is pretty easy.

**Pull request merged**